feat(boot): implement Service booter

bajtos opened the pull request:
Implement automatic discovery and registration of Service providers - see #1439 for the full context and #1649 for the first step on that journey.

This pull request has three parts/commits, the first two are preparations making the actual implementation easier to make and troubleshoot.

**feat(testlab): expose "sandbox.path" property** (4e86b08)

Allow TestSandbox consumers to access the path of the sandbox directory.

Clean up the code by moving "validateInst" check into "path" getter,
thus avoiding the need to explicitly call "validateInst" in every
public API method.

**feat(boot): add debug logs for better troubleshooting** (f88e776)

Make it easier to troubleshoot the situation when a booter is not
recognizing artifact files and/or classes exported by those files.

To make debug logs easy to read, absolute paths are converted to
project-relative paths in debug logs. This change required a refactoring
of Booter design, where "projectRoot" becomes a required constructor
argument.

While making these changes, I changed "options" to be a required
constructor argument too, and chaged both "options" and "projectRoot"

**feat(boot): implement Service booter**

The main change.

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

raymondfeng commented on packages/boot/src/booters/service.booter.ts:
```diff
@@ -0,0 +1,82 @@
+};
+
+function isServiceProvider(cls: Constructor<{}>): cls is ServiceProviderClass {
+  return /Provider$/.test(cls.name);
```
I think we should check the existence of `value()`.

bajtos replied:
Sounds good, I add that check to the existing name check.

(IMO, checking just for `value` is not enough, as there may be other classes/interfaces that contain `value` method too.)

virkt25 replied:
There should be documentation stating that the Booter is expecting `Provider` to be a part of the class name.

raymondfeng commented on packages/boot/test/fixtures/service-class.artifact.ts:
```diff
@@ -0,0 +1,10 @@
+  greet(whom: string = 'world') {
+    return Promise.resolve(`Hello ${whom}`);
+  }
+}
```
Are we going to bind plain service class during boot? I don't see the logic.

bajtos replied:
Not yet. The logic here is to ensure that the booter is picking only Providers and excludes regular classes, there is an integration test for that.

https://github.com/strongloop/loopback-next/blob/9098ad3ade0b226e90291c0f3365b290bd508a66/packages/boot/test/integration/service.booter.integration.ts#L23-L26

Suggestions are welcome on how to make this more obvious to people reading this code.

Maybe it's a sign that I should enhance `ServiceMixin` and the booter to handle both `app.serviceProvider(GeocoderServiceProvider)` and `app.service(GreeterService)` cases, even though `app.service` may not be used by any real LB4 app? IDK.

raymondfeng replied:
For now, let's just add a comment to `GreeterService` saying it will be excluded from discovery as we only allow providers so far.

raymondfeng commented on packages/boot/src/booters/base-artifact.booter.ts:
```diff
@@ -78,7 +95,25 @@ export class BaseArtifactBooter implements Booter {
+
     this.discovered = await discoverFiles(this.glob, this.projectRoot);
+
+    if (debug.enabled) {
```
Because JSON.stringify is expensive. The parameters will be calculated before invoking debug itself.

bajtos commented:
I have addressed your review comments and finished updates of example apps and docs - see 36a8eac.

The pull request is ready for the final review and landing.

@raymondfeng @virkt25 PTAL again.

raymondfeng commented on docs/site/Booting-an-Application.md:
```diff
@@ -237,6 +237,33 @@ The `datasources` object support the following options:
+
+**IMPORTANT:** For a class to be recognized by `ServiceBooter` as a service
+provider, its name must end with `Provider` suffix and its prototype must have a
+`value()` method.
```
To me, this is a sign that we should have something like https://github.com/strongloop/loopback-next/pull/992 to make it possible to explicit.

virkt25 replied:
+1 for #992, it's certainly needed.

raymondfeng commented on docs/site/soap-calculator-tutorial-scaffolding.md:
````diff
@@ -12,12 +12,14 @@ Let's start by creating the initial application by running the following
 
 ```sh
-lb4 app soap-calculator --enableRepository
+lb4 app soap-calculator --enableRepository --enableServices
````
Hmm, now we have `enableRepository` vs `enableServices`. The naming is inconsistent.
